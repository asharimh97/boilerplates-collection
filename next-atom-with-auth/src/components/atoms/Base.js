import styled from "styled-components";
import {
  compose,
  layout,
  position,
  space,
  color,
  flexbox,
  grid,
  background,
  border,
  shadow,
  typography
} from "styled-system";
import propTypes from "@styled-system/prop-types";

import theme from "styles/theme";

const Base = styled("div")(
  compose(
    background,
    border,
    color,
    flexbox,
    grid,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

Base.displayName = "Base";
Base.defaultProps = {
  theme
};

Base.propTypes = {
  ...propTypes.background,
  ...propTypes.border,
  ...propTypes.color,
  ...propTypes.flexbox,
  ...propTypes.grid,
  ...propTypes.layout,
  ...propTypes.position,
  ...propTypes.shadow,
  ...propTypes.space,
  ...propTypes.typography
};

export default Base;
