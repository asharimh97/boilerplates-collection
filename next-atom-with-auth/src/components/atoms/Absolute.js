import styled from "styled-components";
import Base from "./Base";

const Absolute = styled(Base)`
  position: absolute;
`;

export default Absolute;
