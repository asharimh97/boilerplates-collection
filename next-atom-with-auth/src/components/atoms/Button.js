import { Button as AntButton } from "antd";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";
import { flexbox, layout, space, typography } from "styled-system";
import propTypes from "@styled-system/prop-types";

const flex = props => props.flex && { display: "flex" };

const color = props => {
  if (props.color) {
    if (props.type === "link" || props.type === "text") {
      return css`
        &,
        &:hover,
        &:active,
        &:focus {
          color: ${props.theme.colors[props.color]?.["primary"] || props.color};
        }
      `;
    }

    return css`
      background-color: ${props.theme.colors[props.color]?.["primary"] ||
      props.color};
      border-color: ${props.theme.colors[props.color]?.["primary"] ||
      props.color};
      color: white;

      &:hover,
      &:active,
      &:focus {
        background-color: ${props.theme.colors[props.color]?.["60"] ||
        props.color};
        border-color: ${props.theme.colors[props.color]?.["60"] || props.color};
        color: white;
      }
    `;
  }

  return null;
};

const Button = styled(AntButton)`
  ${color};

  ${flex};
  ${flexbox};
  ${layout};
  ${space};
  ${typography}
`;

Button.propTypes = {
  flex: PropTypes.bool,
  ...propTypes.flexbox,
  ...propTypes.layout,
  ...propTypes.space,
  ...propTypes.typography
};

Button.displayName = "Button";

export default Button;
