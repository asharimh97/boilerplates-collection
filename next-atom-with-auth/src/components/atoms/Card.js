import styled from "styled-components";
import Base from "./Base";

const Card = styled(Base)``;

Card.defaultProps = {
  background: "white",
  p: 3,
  borderRadius: "5px",
  boxShadow: 5
};

Card.displayName = "Card";

export default Card;
