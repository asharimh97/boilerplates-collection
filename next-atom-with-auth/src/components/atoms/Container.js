import styled from "styled-components";
import Base from "./Base";

const flex = props => (props.flex ? { display: "flex" } : null);

const fluid = props => props.fluid && { maxWidth: "100%" };

const Container = styled(Base)`
  ${flex};
  ${fluid};
`;

Container.displayName = "Container";

Container.defaultProps = {
  mx: "auto",
  width: "100%",
  maxWidth: 640
};

export default Container;
