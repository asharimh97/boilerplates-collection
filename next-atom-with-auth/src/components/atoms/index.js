export { default as Absolute } from "./Absolute";
export { default as Avatar } from "./Avatar";
export { default as Box } from "./Box";
export { default as Button } from "./Button";
export { default as Card } from "./Card";
export { default as Container } from "./Container";
export { default as Divider } from "./Divider";
export { default as Fixed } from "./Fixed";
export { default as Flex } from "./Flex";
export { default as Grid } from "./Grid";
export { default as Image } from "./Image";
export { default as Link } from "./Link";
export { default as Sticky } from "./Sticky";
export { default as Text } from "./Text";
export { default as Title } from "./Title";
