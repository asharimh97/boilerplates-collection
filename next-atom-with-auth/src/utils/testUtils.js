import { createStore, applyMiddleware } from "redux";

import { middlewares } from "stores";
import reducers from "stores/reducers";
import { shallow } from "enzyme";

import { render as rtlRender } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { create as rtrCreate } from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

export const render = (component, options) => rtlRender(component, options);

export const create = component => rtrCreate(component);

export const fireEvent = userEvent;

/**
 * Function to find a component with `data-test` attribute
 * @param {ShallowWrapper} wrapper - Shallowed component.
 * @param {String} val -  `data-test` attribute value.
 * @return {ShallowWrapper}
 */
export const findByTestAttribute = (wrapper, val) => {
  return wrapper.find(`[data-test="${val}"]`);
};

/**
 * Function to validate/test if component is appear on component
 * @param {ShallowWrapper} wrapper - Shallowed component.
 * @param {String} componentName - component name to look out.
 */
export const findComponent = (wrapper, componentName) => {
  const comp = findByTestAttribute(wrapper, componentName);
  expect(comp.length).toBe(1);
};

/**
 * Function to validate/test if component is appear on component
 * @param {ShallowWrapper} wrapper - Shallowed component.
 * @param {String} componentName - component name to look out based on data-test.
 * @param {Number} total - expected total
 */
export const findComponentTotal = (wrapper, componentName, total) => {
  const comp = findByTestAttribute(wrapper, componentName);
  expect(comp.length).toBe(total);
};

/**
 * Function to validate if component is not found/appear in tested component
 * @param {ShallowWrapper} wrapper - Shallowed component.
 * @param {String} componentName - component name to look out.
 */
export const unFindComponent = (wrapper, componentName) => {
  const comp = findByTestAttribute(wrapper, componentName);
  expect(comp.length).toBe(0);
};

/**
 * Function to validate if action is exist in connected component
 * @param {ShallowWrapper} wrapper - Shallowed component
 * @param {String} actionName - action name
 */
export const findActionProp = (wrapper, actionName) => {
  const actionProp = wrapper.instance().props[actionName];
  expect(actionProp).toBeInstanceOf(Function);
};

/**
 * Create a testing store with imported reducers, middleware, and state
 *  globals: reducers
 * @param {Object} state - State for the store
 * @function storeFactory
 * @return {Store}
 */
export const storeFactory = state => {
  const storeWithMiddleware = createStore(
    reducers,
    state,
    applyMiddleware(...middlewares)
  );
  return storeWithMiddleware;
};

/**
 * Mock next route since next route is meant to be in client side
 * while testing component which uses next route is tested in
 * server side.
 */
export const mockedNextRouter = {
  pushRoute: () => {},
  route: "",
  push: () => {},
  replace: () => {}
};

/**
 * Function to shorten shallowing unconnected component
 * @param {Node} Component - Component being shallowed
 * @param {Object} props - properties
 * @param {Object} state - state given to component
 * @return {ShallowWrapper}
 */
export const shallowComponent = (Component, props = {}, state = null) => {
  const wrapper = shallow(<Component {...props} />);

  if (state) wrapper.setState(state);

  return wrapper;
};

/**
 * Function to shorten shallowing redux-connected component non-page and non-form
 * @param {*} Component - Component being shallowed
 * @param {*} state - redux state value
 * @param {*} props - additional prop
 * @return {ShallowWrapper}
 */
export const shallowReduxComponent = (Component, state = {}, props = {}) => {
  const store = storeFactory(state);
  const wrapper = shallow(<Component store={store} {...props} />)
    .dive()
    .dive();

  return wrapper;
};

/**
 * Function to shorten shallowing redux-connected page/form component
 * @param {*} Component - Component being shallowed
 * @param {*} state - redux state value
 * @param {*} props - additional prop
 * @return {ShallowWrapper}
 */
export const shallowConnect = (Component, state = {}, props = {}) => {
  const store = storeFactory(state);
  const wrapper = shallow(<Component store={store} {...props} />)
    .dive()
    .dive()
    .dive();

  return wrapper;
};
