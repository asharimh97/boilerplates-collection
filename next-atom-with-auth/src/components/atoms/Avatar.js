import styled, { css } from "styled-components";
import { convertNumberToPixel } from "utils/helper";
import Image from "./Image";

const sizes = {
  small: { width: "40px", height: "40px" },
  regular: { width: "50px", height: "50px" },
  big: { width: "65px", height: "65px" }
};

const Avatar = styled(Image).attrs(props => ({
  autofit: true,
  height: sizes[props.size]?.width || convertNumberToPixel(props.size),
  fullRounded: props.circle
}))`
  ${props =>
    props.size && !Number.isInteger(props.size)
      ? css`
          ${sizes[props.size]}
        `
      : css`
          width: ${props.size};
          height: ${props.size};
        `}
`;

Avatar.propTypes = {};

Avatar.displayName = "Avatar";

Avatar.defaultProps = {
  size: "regular",
  src: "/static/images/avatar.jpeg"
};

export default Avatar;
