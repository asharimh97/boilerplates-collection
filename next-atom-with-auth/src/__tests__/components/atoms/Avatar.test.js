import "";
import { Avatar } from "components/atoms";
import { create, render } from "utils/testUtils";

test("renders `Avatar` without error", () => {
  const { getAllByTestId } = render(<Avatar data-testid="avatar-component" />);

  expect(getAllByTestId("avatar-component").length).not.toBe(0);
});

const setup = props => create(<Avatar {...props} />).toJSON();

describe("test styled `Avatar`", () => {
  test("render with default style", () => {
    const avatar = setup();

    expect(avatar).toMatchSnapshot();
    expect(avatar).toHaveStyleRule("width", "50px");
    expect(avatar).toHaveStyleRule("height", "50px");
  });
  test("render with prop `small`", () => {
    const avatar = setup({ size: "small" });

    expect(avatar).toMatchSnapshot();
    expect(avatar).toHaveStyleRule("width", "40px");
    expect(avatar).toHaveStyleRule("height", "40px");
  });
  test("render with prop `big`", () => {
    const avatar = setup({ size: "big" });

    expect(avatar).toMatchSnapshot();
    expect(avatar).toHaveStyleRule("width", "65px");
    expect(avatar).toHaveStyleRule("height", "65px");
  });
  test("render with prop `size`", () => {
    const avatar = setup({ size: 30 });

    expect(avatar).toMatchSnapshot();
    expect(avatar).toHaveStyleRule("width", "30px");
    expect(avatar).toHaveStyleRule("height", "30px");
  });
  test("render with prop `size`", () => {
    const avatar = setup({ size: "1em" });

    expect(avatar).toMatchSnapshot();
    expect(avatar).toHaveStyleRule("width", "1em");
    expect(avatar).toHaveStyleRule("height", "1em");
  });
});
