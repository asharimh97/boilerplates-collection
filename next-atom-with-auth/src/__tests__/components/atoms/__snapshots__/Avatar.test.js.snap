// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test styled \`Avatar\` render with default style 1`] = `
.c0 {
  width: regular;
  height: 50px;
}

.c1 {
  position: relative;
}

.c3 {
  height: 100% !important;
  object-position: center;
  object-fit: cover;
  position: absolute;
  width: 100% !important;
}

.c2 {
  width: 50px;
  height: 50px;
}

<div
  className="c0 c1 c2"
  height="50px"
  size="regular"
>
  <img
    alt=""
    className="c3 c2"
    height="50px"
    onError={[Function]}
    size="regular"
    src="/static/images/avatar.jpeg"
  />
</div>
`;

exports[`test styled \`Avatar\` render with prop \`big\` 1`] = `
.c0 {
  width: big;
  height: 65px;
}

.c1 {
  position: relative;
}

.c3 {
  height: 100% !important;
  object-position: center;
  object-fit: cover;
  position: absolute;
  width: 100% !important;
}

.c2 {
  width: 65px;
  height: 65px;
}

<div
  className="c0 c1 c2"
  height="65px"
  size="big"
>
  <img
    alt=""
    className="c3 c2"
    height="65px"
    onError={[Function]}
    size="big"
    src="/static/images/avatar.jpeg"
  />
</div>
`;

exports[`test styled \`Avatar\` render with prop \`size\` 1`] = `
.c0 {
  width: 30px;
  height: 30px;
}

.c1 {
  position: relative;
}

.c3 {
  height: 100% !important;
  object-position: center;
  object-fit: cover;
  position: absolute;
  width: 100% !important;
}

.c2 {
  width: 30px;
  height: 30px;
}

<div
  className="c0 c1 c2"
  height="30px"
  size={30}
>
  <img
    alt=""
    className="c3 c2"
    height="30px"
    onError={[Function]}
    size={30}
    src="/static/images/avatar.jpeg"
  />
</div>
`;

exports[`test styled \`Avatar\` render with prop \`size\` 2`] = `
.c0 {
  width: 1em;
  height: 1em;
}

.c1 {
  position: relative;
}

.c3 {
  height: 100% !important;
  object-position: center;
  object-fit: cover;
  position: absolute;
  width: 100% !important;
}

.c2 {
  width: 1em;
  height: 1em;
}

<div
  className="c0 c1 c2"
  height="1em"
  size="1em"
>
  <img
    alt=""
    className="c3 c2"
    height="1em"
    onError={[Function]}
    size="1em"
    src="/static/images/avatar.jpeg"
  />
</div>
`;

exports[`test styled \`Avatar\` render with prop \`small\` 1`] = `
.c0 {
  width: small;
  height: 40px;
}

.c1 {
  position: relative;
}

.c3 {
  height: 100% !important;
  object-position: center;
  object-fit: cover;
  position: absolute;
  width: 100% !important;
}

.c2 {
  width: 40px;
  height: 40px;
}

<div
  className="c0 c1 c2"
  height="40px"
  size="small"
>
  <img
    alt=""
    className="c3 c2"
    height="40px"
    onError={[Function]}
    size="small"
    src="/static/images/avatar.jpeg"
  />
</div>
`;
